SSA/Projects/Scf/Scf.lean

```diff
@@ -0,0 +1,456 @@
+    (regArgs := .nil)
+
+def neg {Γ : Ctxt _} (a : Var Γ .int) : Expr Op Γ .int :=
+  Expr.mk
+    (op := .neg)
+    (ty_eq := rfl)
+    (args := .cons a <| .nil)
+    (regArgs := .nil)
+
+def iterate {Γ : Ctxt _} (k : Nat) (input : Var Γ Ty.int) (body : Com Op [.int] .int) : Expr Op Γ .int :=
+  Expr.mk
+    (op := .iterate k)
+    (ty_eq := rfl)
+    (args := .cons input .nil)
+    (regArgs := HVector.cons body HVector.nil)
+
+def if_ {Γ : Ctxt _} {t t': Ty} (cond : Var Γ .bool) (v : Var Γ t) (then_ else_ : Com Op [t] t') : Expr Op Γ t' :=
+  Expr.mk
+    (op := .if t t')
+    (ty_eq := rfl)
+    (args := .cons cond <| .cons v .nil)
+    (regArgs := HVector.cons then_ <| HVector.cons else_ <| HVector.nil)
+
+def run {Γ : Ctxt _} {t : Ty} (v : Var Γ t) (body : Com Op [t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .run t)
+    (ty_eq := rfl)
+    (args := .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+def for_ {Γ : Ctxt _} {t : Ty} (start step : Var Γ .int) (niter : Var Γ .nat) (v : Var Γ t) (body : Com Op [.int, t] t) : Expr Op Γ t :=
+  Expr.mk
+    (op := .for t)
+    (ty_eq := rfl)
+    (args := .cons start <| .cons step <| .cons niter <| .cons v .nil)
+    (regArgs := HVector.cons body <| HVector.nil)
+
+/-- 'if' condition of a true variable evaluates to the then region body. -/
+theorem if_true {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = true) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v then_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+/-- 'if' condition of a false variable evaluates to the else region body. -/
+theorem if_false {t : Ty} (cond : Var Γ .bool) (hcond : Γv cond = false) (v : Var Γ t) (then_ else_ : Com Op [t] t) :
+  Expr.denote (ScfRegion.if_ (t := t) cond v then_ else_) Γv =
+  Expr.denote (ScfRegion.run (t := t) v else_) Γv := by
+    simp[Expr.denote, if_, run]
+    simp_peephole[hcond] at Γv
+    simp[ite_true]
+
+
+/-- a region that returns the value immediately -/
+abbrev RegionRet [Goedel Ty] (t : Ty) {Γ : Ctxt Ty} (v : Var Γ t) : Com Op Γ t := .ret v
+
+/-- a for loop whose body immediately returns the loop variable is the same as
+  just fetching the loop variable. -/
+theorem for_return {t : Ty} (istart istep: Var Γ .int) (niters : Var Γ .nat) (v : Var Γ t) :
+  Expr.denote (ScfRegion.for_ (t := t) istart istep niters v (RegionRet t ⟨1, by simp⟩)) Γv = Γv v := by
+    simp[Expr.denote, run, for_]
+    simp_peephole at Γv
+    simp[Com.denote]
+    rw[loop_counter_decorator_constant_iterate]
+
+
+set_option pp.proofs false
+set_option pp.proofs.withType false
+
+/-# Repeatedly adding a constant in a loop is replaced with a multiplication.
+
+We keep the increment outside the loop so that we don't need to deal with creating and deleting tuples for the "for" region body,
+since our regions are isolatedFromAbove.
```

Comment:
How can we keep the increment outside, if the region is isolated from above?

Reply:
> How can we keep the increment outside, if the region is isolated from above?

The increment variable isn't in the IR, it's a "lean variable". I'll change the proof to bring it into the IR in the second pass.